### PR TITLE
Add [Gen 8] Random Battle (No Dynamax)

### PR DESCRIPTION
**Pull request by ACakeWearingAHat**
This format was polled for in the Random Battles room and received over 1100 votes, an 80% supermajority, in support of its existence. This pull request defines a format with some different sets to Gen 8 Random Battle, but mostly the same. Levels are tier-based. Sets are structured around dynamax not existing. This uses the primary files for Gen 8 Random Battle as a base.

Big thanks to @AnnikaCodes for answering my questions as I figured this out.

to do: level gmax mons without hardcoding each of them individually

### CODE REVIEW

File: data/formats-data.ts [hunk 1]
```diff
@@ -33,6 +33,7 @@ export const FormatsData: {[k: string]: SpeciesFormatsData} = {
 		randomBattleLevel: 82,
 		randomDoubleBattleMoves: ["airslash", "heatwave", "overheat", "protect", "scorchingsands", "tailwind"],
 		randomDoubleBattleLevel: 80,
+		randomBattleNoDynamaxMoves: ["defog", "earthquake", "fireblast", "hurricane", "roost", "toxic"],
```
Comment (KrisXV) [inline data/formats-data.ts:36]: this is getting somewhat ridiculous.... i don't really like containing all random battles elements in formats-data.ts and was aiming on making the file eventually obsolete apart from tiering.

File: data/random-teams.ts [hunk 2]
```diff
@@ -1056,6 +1058,8 @@ export class RandomTeams {
 			return {cull: pulseIncompatible && !shiftryCase && counter.setupType !== 'Special'};
 		case 'suckerpunch':
 			return {cull:
+				// Shiftry in No Dyna would otherwise get Choice Scarf Sucker Punch sometimes.
```
Comment (KrisXV) [inline data/random-teams.ts:1061]: please do not shorten it to "dyna"

File: data/formats-data.ts (hunk 1 above)
Comment (AnnikaCodes) [reply data/formats-data.ts:36]: I don't have time to review this PR tonight but I have plans to move random battles data (including selection logic for moves/item/abilities/etc) into its own data file and just have one random-teams.ts file. That would solve the issue of clutter in formats-data as well as separation of code and data in randbats (making it easier for inexperienced people to understand the logic and cleaning things up a lot for experienced programmers). I'm unsure of when I'll have time to work on that though, perhaps this summer?

File: data/random-teams.ts [hunk 3]
```diff
@@ -1011,7 +1012,8 @@ export class RandomTeams {
 			return {cull: hasMove['dragondance']};
 		case 'hurricane':
 			// Special case for Noctowl, which wants Air Slash if Nasty Plot instead
-			return {cull: counter.setupType === 'Physical' || (hasAbility['Tinted Lens'] && counter.setupType && !isDoubles)};
+			const noctowlCase = (!isNoDyna && hasAbility['Tinted Lens'] && counter.setupType && !isDoubles);
```
Comment (AnnikaCodes) [inline data/random-teams.ts:1015]:
```suggestion
			const noctowlCase = (!isNoDyna && !isDoubles && species.id === 'noctowl' && counter.setupType);
```
This rejection seems to be unrelated to the actual Tinted Lens ability.

File: data/random-teams.ts [hunk 4]
```diff
@@ -1237,20 +1243,33 @@ export class RandomTeams {
 			return hasMove['dragondance'];
 		case 'Sheer Force':
 			return (!counter.sheerforce || hasAbility['Guts']);
+		case 'Shell Armor':
+			return (species.id === 'omastar' && (hasMove['spikes'] || hasMove['stealthrock']));
```
Comment (AnnikaCodes) [inline data/random-teams.ts:1247]: Can we split the changes that affect the main format into their own PR/commit?

File: data/random-teams.ts [hunk 5]
```diff
@@ -1265,7 +1284,10 @@ export class RandomTeams {
 				movePool.includes('snarl')
 			);
 		case 'Tinted Lens':
-			return (hasMove['defog'] || hasMove['hurricane'] || counter.Status > 2 && !counter.setupType);
+			return (hasMove['defog'] || (!isNoDyna && hasMove['hurricane']) ||
+			(counter.Status > 2 && !counter.setupType) ||
+			(isNoDyna && species.id === 'butterfree')
+			);
```
Comment (AnnikaCodes) [inline data/random-teams.ts:1290]:
```suggestion
			return (
				hasMove['defog'] ||
				(!isNoDyna && hasMove['hurricane']) ||
				(counter.Status > 2 && !counter.setupType) ||
				(isNoDyna && species.id === 'butterfree')
			);
```

File: data/random-teams.ts [hunk 6]
```diff
@@ -1524,7 +1550,9 @@ export class RandomTeams {
 			ability !== 'Sturdy' &&
 			(species.baseStats.spe >= 90 || !hasMove['voltswitch']) &&
 			['foulplay', 'rapidspin', 'substitute', 'uturn'].every(m => !hasMove[m]) && (
-				counter.speedsetup || counter.drain ||
+				counter.speedsetup ||
+				// No Dyna Buzzwole doesn't want Life Orb with Bulk Up + 3 attacks
```
Comment (AnnikaCodes) [inline data/random-teams.ts:1554]: You say this applies only to No Dyna Buzzwole, but never check if we're in the normal or no-Dynamax format.

File: data/random-teams.ts [hunk 7]
```diff
@@ -1788,7 +1820,41 @@ export class RandomTeams {
 			forme = 'Pikachu' + this.sample(['', '-Original', '-Hoenn', '-Sinnoh', '-Unova', '-Kalos', '-Alola', '-Partner', '-World']);
 		}
 
-		const level: number = (isDoubles ? species.randomDoubleBattleLevel : species.randomBattleLevel) || 80;
+		let level: number;
+		if (isDoubles && species.randomDoubleBattleLevel) {
+			level = species.randomDoubleBattleLevel;
+		} else if (isNoDyna) {
+			const tierScale: {[k: string]: number} = {
```
Comment (AnnikaCodes) [inline data/random-teams.ts:1827]:
```suggestion
			const tier = species.name.endsWith('-Gmax') ? this.dex.getSpecies(species.changesFrom).tier : species.tier;
			const tierScale: {[k: string]: number} = {
```

File: data/random-teams.ts [hunk 8]
```diff
@@ -1788,7 +1820,41 @@ export class RandomTeams {
...
+				// These Pokemon are too strong and need a lower level
+				zaciancrowned: 66, calyrexshadow: 68, xerneas: 70, necrozmaduskmane: 72, glalie: 78, octillery: 84, wobbuffet: 86,
+				// These Pokemon are too weak and need a higher level
+				delibird: 100, vespiquen: 96, pikachu: 92,
+				// G-max formes are AG so i have to tier these on their own, fix this in review if you can
+				blastoisegmax: 86, butterfreegmax: 90, gengargmax: 81, laprasgmax: 90, snorlaxgmax: 84, garbodorgmax: 90,
+				rillaboomgmax: 80, inteleongmax: 86, coalossalgmax: 90, hatterenegmax: 82, grimmsnarlgmax: 82, alcremiegmax: 90,
+				copperajahgmax: 86, urshifugmax: 76, urshifurapidstrikegmax: 80,
+			};
+			level = customScale[species.id] || tierScale[species.tier];
```
Comment (AnnikaCodes) [inline data/random-teams.ts:1851]:
```suggestion
			};
			level = customScale[species.id] || tierScale[tier];
```

File: sim/dex-species.ts [hunk 9]
```diff
@@ -32,6 +32,7 @@ export interface SpeciesFormatsData {
 	randomBattleLevel?: number;
 	randomDoubleBattleMoves?: readonly string[];
 	randomDoubleBattleLevel?: number;
+	randomBattleNoDynamaxMoves?: readonly string [];
```
Comment (AnnikaCodes) [inline sim/dex-species.ts:35]:
```suggestion
	randomBattleNoDynamaxMoves?: readonly string[];
```

**Review by AnnikaCodes — CHANGES_REQUESTED** [review]

File: data/random-teams.ts (hunk 4 above)
Comment (AnnikaCodes) [reply data/random-teams.ts:1247]: oh, I'm told that Omastar never gets shell armor in other formats

File: data/random-teams.ts [hunk 10]
```diff
@@ -885,7 +886,7 @@ export class RandomTeams {
 		case 'stickyweb':
 			return {cull: counter.setupType === 'Special' || !!teamDetails.stickyWeb};
 		case 'taunt':
-			return {cull: hasMove['nastyplot'] || hasMove['swordsdance']};
+			return {cull: hasMove['encore'] || hasMove['nastyplot'] || hasMove['swordsdance']};
```
[the remark this answers is not in the document]
Comment (AnnikaCodes) [reply data/random-teams.ts:889]: (resolved in PMs)

File: data/formats-data.ts [hunk 11]
```diff
@@ -249,10 +250,11 @@ export const FormatsData: {[k: string]: SpeciesFormatsData} = {
 		doublesTier: "(DUU)",
 	},
 	raichu: {
-		randomBattleMoves: ["surf", "focusblast", "grassknot", "nastyplot", "thunderbolt", "voltswitch"],
+		randomBattleMoves: ["focusblast", "grassknot", "nastyplot", "surf", "thunderbolt", "voltswitch"],
```
[the remark this answers is not in the document]
Comment (ACakeWearingAHat) [reply data/formats-data.ts:253]: alphabetization

File: data/random-teams.ts (hunk 4 above)
Comment (ACakeWearingAHat) [reply data/random-teams.ts:1247]: ya

**Review by AnnikaCodes — APPROVED** [review]
Alright, let's do this!
